File: src/data/categorias.ts
import type { Categoria } from '@/types';

// Escala oficial de Monotributo vigente (ARCA). Verificado 2026-06 contra la tabla de categorías de
// ARCA (arca.gob.ar/monotributo/categorias.asp) y el facturómetro del propio contribuyente.
//   topeAnual               = "Ingresos brutos" (límite anual de la categoría).
//   cuotaServicios/Comercio = "Total" de cada actividad (impuesto integrado + aportes SIPA + obra
//                             social); es el FALLBACK de la cuota mensual cuando no tenemos el
//                             importe real del cliente (cliente.proxVencImporte lo pisa).
//   alquilerMaxAnual        = "Alquileres devengados anualmente".
//   topePrecioUnitario      = "Precio unitario máximo para venta de cosas muebles".
// OJO: ARCA reajusta esta escala cada semestre. Estos valores son el FALLBACK: al iniciar, el front
// pide la escala vigente al backend (GET /indicadores/categorias, tabla pública de ARCA) y la pisa
// con `aplicarMontosOficiales`. Si esa fuente falla, quedan estos valores. Mantener actualizados igual.
export const CATEGORIAS: Categoria[] = [
  { codigo: 'A', topeAnual: 10_277_988,  cuotaServicios: 42_387,    cuotaComercio: 42_387,   superficieMax: 30,  energiaMaxKwh: 3_330,   alquilerMaxAnual: 2_390_230 },
  { codigo: 'B', topeAnual: 15_058_448,  cuotaServicios: 48_251,    cuotaComercio: 48_251,   superficieMax: 45,  energiaMaxKwh: 5_000,   alquilerMaxAnual: 2_390_230 },
  { codigo: 'C', topeAnual: 21_113_697,  cuotaServicios: 56_502,    cuotaComercio: 55_227,   superficieMax: 60,  energiaMaxKwh: 6_700,   alquilerMaxAnual: 3_266_647 },
  { codigo: 'D', topeAnual: 26_212_853,  cuotaServicios: 72_414,    cuotaComercio: 70_661,   superficieMax: 85,  energiaMaxKwh: 10_000,  alquilerMaxAnual: 3_266_647 },
  { codigo: 'E', topeAnual: 30_833_964,  cuotaServicios: 102_538,   cuotaComercio: 92_658,   superficieMax: 110, energiaMaxKwh: 13_000,  alquilerMaxAnual: 4_143_065 },
  { codigo: 'F', topeAnual: 38_642_048,  cuotaServicios: 129_045,   cuotaComercio: 111_198,  superficieMax: 150, energiaMaxKwh: 16_500,  alquilerMaxAnual: 4_143_065 },
  { codigo: 'G', topeAnual: 46_211_109,  cuotaServicios: 197_108,   cuotaComercio: 135_918,  superficieMax: 200, energiaMaxKwh: 20_000,  alquilerMaxAnual: 4_939_808 },
  { codigo: 'H', topeAnual: 70_113_407,  cuotaServicios: 447_347,   cuotaComercio: 272_063,  superficieMax: 200, energiaMaxKwh: 20_000,  alquilerMaxAnual: 7_170_689 },
  { codigo: 'I', topeAnual: 78_479_212,  cuotaServicios: 824_802,   cuotaComercio: 406_512,  superficieMax: 200, energiaMaxKwh: 20_000,  alquilerMaxAnual: 7_170_689, topePrecioUnitario: 613_492 },
  { codigo: 'J', topeAnual: 89_872_640,  cuotaServicios: 999_008,   cuotaComercio: 497_059,  superficieMax: 200, energiaMaxKwh: 20_000,  alquilerMaxAnual: 7_170_689, topePrecioUnitario: 613_492 },
  { codigo: 'K', topeAnual: 108_357_084, cuotaServicios: 1_381_688, cuotaComercio: 600_880,  superficieMax: 200, energiaMaxKwh: 20_000,  alquilerMaxAnual: 7_170_689, topePrecioUnitario: 613_492 },
];

// `let` (no const) a propósito: `aplicarMontosOficiales` lo repisa con el valor vigente de ARCA. Los
// imports son bindings VIVOS (ESM), así que los consumidores ven el valor actualizado.
export let TOPE_CATEGORIA_K = 108_357_084;

// Precio unitario máximo para la venta de cosas muebles (productos). A diferencia de los topes de
// ingresos, este límite es ÚNICO y aplica a TODAS las categorías (A→K), no sólo a las altas: un
// monotributista no puede vender un producto a un precio unitario mayor a este valor. No alcanza a
// servicios. ARCA no rechaza el comprobante que lo supera, pero queda registrado → lo avisamos al
// emitir. (El campo `topePrecioUnitario` por categoría queda como referencia del dato oficial.)
export let TOPE_PRECIO_UNITARIO = 613_492;

// Pisa la escala local con los montos OFICIALES vigentes (tabla pública de ARCA, vía el backend).
// Muta CATEGORIAS in-place (los imports guardan la MISMA referencia de array, así que ven los nuevos
// valores) y reasigna los topes derivados. Si nunca se llama (backend caído), quedan los valores
// hardcodeados como fallback. Se llama una vez al iniciar (ConfigContext).
export function aplicarMontosOficiales(oficiales: readonly Categoria[]): void {
  for (const o of oficiales) {
    const local = CATEGORIAS.find(c => c.codigo === o.codigo);
    if (!local) continue;
    local.topeAnual = o.topeAnual;
    local.cuotaServicios = o.cuotaServicios;
    local.cuotaComercio = o.cuotaComercio;
    local.superficieMax = o.superficieMax;
    local.energiaMaxKwh = o.energiaMaxKwh;
    local.alquilerMaxAnual = o.alquilerMaxAnual;
    if (local.topePrecioUnitario !== undefined && o.topePrecioUnitario) {
      local.topePrecioUnitario = o.topePrecioUnitario;
    }
  }
  const k = CATEGORIAS.find(c => c.codigo === 'K');
  if (k) TOPE_CATEGORIA_K = k.topeAnual;
  const pu = oficiales.find(o => o.topePrecioUnitario)?.topePrecioUnitario;
  if (pu) TOPE_PRECIO_UNITARIO = pu;
}

export function getCategoria(codigo: string | null | undefined): Categoria {
  return CATEGORIAS.find(c => c.codigo === codigo) || CATEGORIAS[0];
}

export function siguienteCategoria(codigo: string): Categoria | null {
  const idx = CATEGORIAS.findIndex(c => c.codigo === codigo);
  if (idx === -1 || idx === CATEGORIAS.length - 1) return null;
  return CATEGORIAS[idx + 1];
}

// Umbrales legales del ratio gastos / tope Cat K (art. 20, inc. j)
export const RATIO_GASTOS_COMERCIO = 0.80;
export const RATIO_GASTOS_SERVICIOS = 0.40;

File: backend/app/data/categorias.py
"""Escala oficial de Monotributo (ARCA), ESPEJO de `src/data/categorias.ts`.

Esta tabla la usa el motor de alertas del backend (services/monotributo.py) para evaluar % del tope,
categoría que corresponde y ratio de gastos sin depender del front.

Los valores de acá son el FALLBACK. La escala se reajusta cada semestre, así que igual que en el
front la tabla se pisa en memoria con la vigente (`aplicar_montos_oficiales`, alimentada por
services/categorias_afip). Que las dos puntas se actualicen solas importa: el front muestra el tope
en pantalla y el backend decide qué alerta se ENVÍA — si una escala se atrasa respecto de la otra,
el contador ve "todo bien" y le llega un aviso de recategorización (o al revés).
"""
from __future__ import annotations

from dataclasses import dataclass


@dataclass(frozen=True)
class Categoria:
    codigo: str
    tope_anual: float
    cuota_servicios: float
    cuota_comercio: float


# Escala de referencia (última verificada: 2026-08). Sólo replicamos los campos que el motor de
# alertas necesita (código + tope + cuotas); superficie/energía/alquiler/precio unitario no entran en
# ninguna alerta. La lista se muta IN-PLACE al aplicar los montos oficiales: quien la importó
# conserva la misma referencia y ve los valores nuevos.
CATEGORIAS: list[Categoria] = [
    Categoria("A", 12_009_410.45, 49_527.18, 49_527.18),
    Categoria("B", 17_595_182.74, 56_379.08, 56_379.08),
    Categoria("C", 24_670_494.31, 66_020.12, 64_530.58),
    Categoria("D", 30_628_651.43, 84_612.93, 82_564.81),
    Categoria("E", 36_028_231.33, 119_811.45, 108_267.51),
    Categoria("F", 45_151_659.41, 150_784.21, 129_930.65),
    Categoria("G", 53_995_798.87, 230_312.94, 158_815.05),
    Categoria("H", 81_924_660.37, 522_706.68, 317_895.01),
    Categoria("I", 91_699_761.90, 963_747.86, 474_992.78),
    Categoria("J", 105_012_519.20, 1_167_299.76, 580_793.69),
    Categoria("K", 126_610_838.75, 1_614_446.04, 702_103.24),
]

# Umbrales legales del ratio gastos / tope Cat K (art. 20, inc. j).
RATIO_GASTOS_COMERCIO = 0.80
RATIO_GASTOS_SERVICIOS = 0.40


def tope_categoria_k() -> float:
    """Tope de la categoría más alta, que es contra el que se mide el ratio de gastos. Es una
    FUNCIÓN y no una constante a propósito: un `from ... import TOPE_CATEGORIA_K` congela el número
    del import y se perdería la actualización de la escala."""
    return CATEGORIAS[-1].tope_anual


def aplicar_montos_oficiales(oficiales) -> bool:
    """Pisa la escala local con la vigente que publica el organismo. Muta CATEGORIAS in-place (misma
    referencia de lista para todos los que la importaron). `oficiales` son CategoriaOficial de
    services/categorias_afip. Devuelve si aplicó algo."""
    if not oficiales:
        return False
    por_codigo = {c.codigo: c for c in oficiales}
    aplicados = 0
    for i, local in enumerate(CATEGORIAS):
        o = por_codigo.get(local.codigo)
        if o is None:
            continue
        CATEGORIAS[i] = Categoria(
            local.codigo, float(o.topeAnual), float(o.cuotaServicios), float(o.cuotaComercio)
        )
        aplicados += 1
    return aplicados > 0


def get_categoria(codigo: str | None) -> Categoria:
    """La categoría del código dado; si es desconocido/None, la más baja (A), igual que el front."""
    for c in CATEGORIAS:
        if c.codigo == codigo:
            return c
    return CATEGORIAS[0]


def inferir_categoria(facturacion_12m: float) -> str:
    """Código de la categoría que encuadra esa facturación 12m (la última si la supera toda).
    Espejo de inferirCategoria() en src/services/clientesService.ts."""
    for c in CATEGORIAS:
        if facturacion_12m <= c.tope_anual:
            return c.codigo
    return CATEGORIAS[-1].codigo
